### PR TITLE
Scripts for deploying a Jenkins VM and a Docker Build pipeline

### PR DESCRIPTION
**Pull request by clguiman**
* Added a groovy script for a basic Jenkins pipeline
* Added a script that installs Jenkins on a linux VM (used by quickstart template)
* Added a script that can add a Jenkins job to a remote server (also included the templates)

### CODE REVIEW

File: jenkins/install-jenkins-and-docker-build-pipeline.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+wget -q -O - https://pkg.jenkins.io/debian/jenkins-ci.org.key | sudo apt-key add -
+sudo sh -c 'echo deb http://pkg.jenkins.io/debian-stable binary/ > /etc/apt/sources.list.d/jenkins.list'
+curl -fsSL https://apt.dockerproject.org/gpg | sudo apt-key add -
+sudo add-apt-repository \
+       "deb https://apt.dockerproject.org/repo/ \
+       ubuntu-$(lsb_release -cs) \
+       main"
+sudo apt-get update --yes
+sudo apt-get install jenkins --yes
+sudo apt-get install jenkins --yes
```
Comment (ejizba) [inline jenkins/install-jenkins-and-docker-build-pipeline.sh]: Remove duplicate 'install jenkins' line

File: jenkins/install-jenkins-and-docker-build-pipeline.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+sudo apt-get install git --yes
+sudo apt-get -y install docker-engine --yes
+sudo gpasswd -a jenkins docker
+sudo service docker restart
+skill -KILL -u jenkins
+sudo service jenkins restart
+
+if [[ "${include_pipeline}" != "Include" ]]
```
Comment (ejizba) [inline jenkins/install-jenkins-and-docker-build-pipeline.sh]: If we really expect people to reuse these scripts outside of the quickstart, we should change the 'include_pipeline' to a flag or boolean. If you're calling this script directly from azuredeploy.json, I think you can change the commandToExecute to be something like this:

./test.sh $(expr "${x}" == "foo")

File: jenkins/add-docker-build-job.sh [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    echo "                                 <git url>"
+    echo "                                 <container registry url>"
+    echo "                                 <container registry user name>"
+    echo "                                 <container registry password>"
+    echo "                                 [opt: credentials id]"
+    echo "                                 [opt: credentials description]"
+    echo "                                 [opt: job short name]"
+    echo "                                 [opt: job display name]"
+    echo "                                 [opt: job description]"
+    echo "                                 [opt: container tag prefix]"
```
Comment (ejizba) [inline jenkins/add-docker-build-job.sh]: The README says it "must have" the container tag prefix, but here it's optional. We should be consistent either way

File: README.md [hunk 4]
```diff
@@ -37,3 +37,21 @@ Migrates an image from the classic image model to the new Azure Resource Manager
...
+## [Jenkins groovy scripts](groovy/)
+> [groovy/basic-docker-build.groovy](groovy/basic-docker-build.groovy)
+
+Sample Jenkins pipeline that clones a git repository, builds the docker image defined in the Docker file and pushes that image to a private container registry.
+The Jenkins Job that uses this groovy script must have these parameters defined:
+
+| Jenkins job parameters  | Description                                                                                                 |
+|-------------------------|-------------------------------------------------------------------------------------------------------------|
+| git_repo                | A public git repository that has a Dockerfile                                                               |
+| docker_tag_prefix       | The image tag prefix (the tag will be in this format: "<prefix>:build_number")                              |
```
Comment (ejizba) [inline README.md]: The prefix note isn't getting displayed. It just looks like ":build_number"

File: jenkins/add-docker-build-job.sh [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+then
+    job_display_name="Basic Docker Build"
+fi
+if [ -z "$job_description" ]
+then
+    job_description="A basic pipeline that builds a Docker container. The job expects a dockerfile inside the git repository"
+fi
+if [ -z "$container_tag_prefix" ]
+then
+    container_tag_prefix="jenkins/myfirstapp"
```
Comment (ejizba) [inline jenkins/add-docker-build-job.sh]: The first part is usually the orgainization or user name, so I would default this to "${jenkins_user}/myfirstapp"

File: jenkins/add-docker-build-job.sh [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+then
+    job_name="basic-docker-build"
+fi
+if [ -z "$job_display_name" ]
+then
+    job_display_name="Basic Docker Build"
+fi
+if [ -z "$job_description" ]
+then
+    job_description="A basic pipeline that builds a Docker container. The job expects a dockerfile inside the git repository"
```
Comment (ejizba) [inline jenkins/add-docker-build-job.sh]: Does the dockerfile have to be at the root of the repo? If so I would say "The job expects a dockerfile at the root of the git repository"

File: jenkins/add-docker-build-job.sh [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+fi
+if [ -z "$container_tag_prefix" ]
+then
+    container_tag_prefix="jenkins/myfirstapp"
+fi
+
+#download dependencies
+base_remote_jenkins_scripts="https://raw.githubusercontent.com/Azure/azure-devops-utils/master/jenkins"
```
Comment (ejizba) [inline jenkins/add-docker-build-job.sh]: Being nitpicky, I would prefer:
base_remote_jenkins="https://raw.githubusercontent.com/Azure/azure-devops-utils/master/"

wget ${base_remote_jenkins}jenkins/basic-docker-build-job.xml -O job_template.xml
wget ${base_remote_jenkins}jenkins/basic-user-pwd-credentials.xml -O credentials_template.xml
wget ${base_remote_jenkins}groovy/basic-docker-build.groovy -O job.groovy

File: jenkins/add-docker-build-job.sh [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+rm job_template.xml
+rm job1.xml
+rm job.groovy
+set +e
+
+function retry_until_successful {
+    ${@}
+    while [ $? -ne 0 ]; do
+        sleep 5
+        ${@}
```
Comment (ejizba) [inline jenkins/add-docker-build-job.sh]: Add a timeout, even if it's a long one

File: jenkins/add-docker-build-job.sh [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+#wait for instance to be back online
+retry_until_successful java -jar jenkins-cli.jar -s ${jenkins_url} version --username ${jenkins_user} --password ${jenkins_password}
+
+#add user/pwd
+retry_until_successful cat credentials.xml | java -jar jenkins-cli.jar -s ${jenkins_url} create-credentials-by-xml SystemCredentialsProvider::SystemContextResolver::jenkins "(global)" --username ${jenkins_user} --password ${jenkins_password}
+#add job
+retry_until_successful cat job.xml | java -jar jenkins-cli.jar -s ${jenkins_url} create-job ${job_name} --username ${jenkins_user} --password ${jenkins_password}
+
+#cleanup
+rm credentials.xml
```
Comment (ejizba) [inline jenkins/add-docker-build-job.sh:120]: Are you intentionally leaving 'job.groovy', 'job_template.xml' and 'credentials_template.xml'?

**Review by ejizba — APPROVED** [review]
No major issues.

Everywhere you mention 'docker tag prefix', it should be 'docker repository'. The tag is technically only the part after the ':' and I think the repository should be required

File: jenkins/install-jenkins-and-docker-build-pipeline.sh (hunk 1 above)
Comment (clguiman) [reply jenkins/install-jenkins-and-docker-build-pipeline.sh]: it's not a duplicate. Some times the first one fails :)

File: jenkins/add-docker-build-job.sh (hunk 9 above)
Comment (clguiman) [reply jenkins/add-docker-build-job.sh:120]: I'm removing them earlier

File: jenkins/install-jenkins-and-docker-build-pipeline.sh (hunk 2 above)
Comment (clguiman) [reply jenkins/install-jenkins-and-docker-build-pipeline.sh]: makes sense. I'll change it

File: jenkins/add-docker-build-job.sh (hunk 3 above)
Comment (clguiman) [reply jenkins/add-docker-build-job.sh]: that's for the groovy script, not the shell script

**Comment by clguiman** [comment]
@EricJizbaMSFT I have fixed the issues, please have another look

**Pull request merged**